File: moderate/lca.rb
parent = { 3 => 8, 8 => 30, 10 => 20, 20 => 8, 29 => 20, 30 => nil, 52 => 30 }

File.open(ARGV[0]).each_line do |line|
  n = line.split.map(&:to_i)
  found = false
  while n[0] && n[0] != n[1]
    i = n[1]
    while i
      if n[0] == i
        found = true
        break
      end
      i = parent[i]
    end
    break if found
    n[0] = parent[n[0]]
  end
  puts n[0]
end

File: moderate/lca.lua
parent = {[3]=8, [8]=30, [10]=20, [20]=8, [29]=20, [30]=nil, [52]=30}

for line in io.lines(arg[1]) do
  local n1, n2 = line:match("(%d+) (%d+)")
  n1, n2 = tonumber(n1), tonumber(n2)
  local found = false
  while n1 and n1 ~= n2 do
    local i = n2
    while i do
      if n1 == i then
        found = true
        break
      end
      i = parent[i]
    end
    if found then break end
    n1 = parent[n1]
  end
  print(n1)
end
